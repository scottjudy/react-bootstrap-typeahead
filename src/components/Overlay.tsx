import type { ModifierArguments, Placement, Options } from '@popperjs/core';
import PropTypes from 'prop-types';
import { CSSProperties, ReactElement, Ref, useState } from 'react';
import { usePopper } from 'react-popper';

import { noop } from '../utils';

const alignValues = ['justify', 'left', 'right'] as const;
export type Align = typeof alignValues[number];

// `Element` is not defined during server-side rendering, so shim it here.
/* istanbul ignore next */
const SafeElement = typeof Element === 'undefined' ? noop : Element;

const propTypes = {
  /**
   * Specify menu alignment. The default value is `justify`, which makes the
   * menu as wide as the input and truncates long values. Specifying `left`
   * or `right` will align the menu to that side and the width will be
   * determined by the length of menu item values.
   */
  align: PropTypes.oneOf(alignValues),
  children: PropTypes.func.isRequired,
  /**
   * Specify whether the menu should appear above the input.
   */
  dropup: PropTypes.bool,
  /**
   * Whether or not to automatically adjust the position of the menu when it
   * reaches the viewport boundaries.
   */
  flip: PropTypes.bool,
  isMenuShown: PropTypes.bool,
  positionFixed: PropTypes.bool,
  // @ts-ignore
  referenceElement: PropTypes.instanceOf(SafeElement),
};

const defaultProps = {
  align: 'justify',
  dropup: false,
  flip: false,
  isMenuShown: false,
  positionFixed: false,
};

export interface OverlayRenderProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  innerRef: Ref<any>;
  inputHeight: number;
  scheduleUpdate: (() => void) | null;
  style: CSSProperties;
}

export interface OverlayProps {
  align: Align;
  children: (props: OverlayRenderProps) => ReactElement | null;
  dropup: boolean;
  flip: boolean;
  isMenuShown: boolean;
  positionFixed: boolean;
  referenceElement?: HTMLElement;
}

const setPopperWidth = {
  enabled: true,
  fn: (data: ModifierArguments<Options>) => {
    // eslint-disable-next-line no-param-reassign
    data.state.styles.popper.width = `${data.state.rects.reference.width}px`;
  },
  name: 'setPopperWidth',
  phase: 'write',
};

export function getModifiers(props: Pick<OverlayProps, 'align' | 'flip'>) {
  const modifiers = [
    {
      enabled: !!props.flip,
      name: 'flip',
    },
  ];

  if (props.align !== 'right' && props.align !== 'left') {
    modifiers.push(setPopperWidth);
  }

  return modifiers;
}

export function getPlacement(
  props: Pick<OverlayProps, 'align' | 'dropup'>
): Placement {
  const x = props.align === 'right' ? 'end' : 'start';
  const y = props.dropup ? 'top' : 'bottom';

  return `${y}-${x}`;
}

const Overlay = ({ referenceElement, ...props }: OverlayProps) => {
  const [popperElement, attachRef] = useState<HTMLElement>();
  const { attributes, styles, forceUpdate } = usePopper(
    referenceElement,
    popperElement,
    {
      modifiers: getModifiers(props),
      placement: getPlacement(props),
      strategy: props.positionFixed ? 'fixed' : 'absolute',
    }
  );

  if (!props.isMenuShown) {
    return null;
  }

  return props.children({
    ...attributes.popper,
    innerRef: attachRef,
    inputHeight: referenceElement?.offsetHeight || 0,
    scheduleUpdate: forceUpdate,
    style: styles.popper,
  });
};

Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;

export default Overlay;
