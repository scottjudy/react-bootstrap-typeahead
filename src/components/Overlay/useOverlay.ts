import type { ModifierArguments, Placement, Options } from '@popperjs/core';
import { useEffect, useState } from 'react';
import { usePopper } from 'react-popper';

import { Align } from '../../types';

const setPopperWidth = {
  enabled: true,
  fn: (data: ModifierArguments<Options>) => {
    // eslint-disable-next-line no-param-reassign
    data.state.styles.popper.width = `${data.state.rects.reference.width}px`;
  },
  name: 'setPopperWidth',
  phase: 'write',
};

export type ReferenceElement = HTMLElement | null;

export interface OverlayOptions {
  align?: Align;
  dropup?: boolean;
  flip?: boolean;
  positionFixed?: boolean;
}

export function getModifiers(props: Pick<OverlayOptions, 'align' | 'flip'>) {
  const modifiers = [
    {
      enabled: !!props.flip,
      name: 'flip',
    },
  ];

  if (props.align !== 'right' && props.align !== 'left') {
    modifiers.push(setPopperWidth);
  }

  return modifiers;
}

export function getPlacement(
  props: Pick<OverlayOptions, 'align' | 'dropup'>
): Placement {
  const x = props.align === 'right' ? 'end' : 'start';
  const y = props.dropup ? 'top' : 'bottom';

  return `${y}-${x}`;
}

export function useOverlay(
  referenceElement: ReferenceElement,
  options: OverlayOptions
) {
  const [popperElement, attachRef] = useState<ReferenceElement>(null);
  const { attributes, styles, forceUpdate } = usePopper(
    referenceElement,
    popperElement,
    {
      modifiers: getModifiers(options),
      placement: getPlacement(options),
      strategy: options.positionFixed ? 'fixed' : 'absolute',
    }
  );

  const refElementHeight = referenceElement?.offsetHeight;

  // Re-position the popper if the height of the reference element changes.
  // Exclude `forceUpdate` from dependencies since it changes with each render.
  useEffect(() => {
    forceUpdate && forceUpdate();
  }, [refElementHeight]); // eslint-disable-line

  return {
    ...attributes.popper,
    // Keep for backwards compatibility.
    innerRef: attachRef,
    overlayRef: attachRef,
    style: styles.popper,
  };
}

export default useOverlay;
