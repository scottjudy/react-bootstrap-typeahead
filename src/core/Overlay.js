// @flow

/* eslint-disable react/no-unused-prop-types */

import * as React from 'react';
import PropTypes from 'prop-types';
import { Popper, type PopperChildrenProps } from 'react-popper';

import { values } from '../utils';
import { ALIGN } from '../constants';

import type { OverlayProps } from '../types';

// `Element` is not defined during server-side rendering, so shim it here.
const SafeElement = typeof Element === 'undefined' ? () => {} : Element;

const propTypes = {
  /**
   * Specify menu alignment. The default value is `justify`, which makes the
   * menu as wide as the input and truncates long values. Specifying `left`
   * or `right` will align the menu to that side and the width will be
   * determined by the length of menu item values.
   */
  align: PropTypes.oneOf(values(ALIGN)),
  children: PropTypes.func.isRequired,
  /**
   * Specify whether the menu should appear above the input.
   */
  dropup: PropTypes.bool,
  /**
   * Whether or not to automatically adjust the position of the menu when it
   * reaches the viewport boundaries.
   */
  flip: PropTypes.bool,
  isMenuShown: PropTypes.bool,
  positionFixed: PropTypes.bool,
  referenceElement: PropTypes.instanceOf(SafeElement),
};

const defaultProps = {
  align: ALIGN.JUSTIFY,
  dropup: false,
  flip: false,
  isMenuShown: false,
  positionFixed: false,
};

function getModifiers({ align, flip }: OverlayProps) {
  return {
    computeStyles: {
      enabled: true,
      fn: ({ styles, ...data }) => ({
        ...data,
        styles: {
          ...styles,
          // Use the following condition instead of `align === 'justify'`
          // since it allows the component to fall back to justifying the
          // menu width if `align` is undefined.
          width: align !== ALIGN.RIGHT && align !== ALIGN.LEFT ?
            // Set the popper width to match the target width.
            data.offsets.reference.width :
            styles.width,
        },
      }),
    },
    flip: {
      enabled: flip,
    },
    preventOverflow: {
      escapeWithReference: true,
    },
  };
}

// Flow expects a string literal value for `placement`.
const PLACEMENT = {
  bottom: {
    end: 'bottom-end',
    start: 'bottom-start',
  },
  top: {
    end: 'top-end',
    start: 'top-start',
  },
};

export function getPlacement({ align, dropup }: OverlayProps) {
  const x = align === ALIGN.RIGHT ? 'end' : 'start';
  const y = dropup ? 'top' : 'bottom';

  return PLACEMENT[y][x];
}

const Overlay = (props: OverlayProps) => {
  const { children, isMenuShown, positionFixed, referenceElement } = props;

  if (!isMenuShown) {
    return null;
  }

  return (
    <Popper
      modifiers={getModifiers(props)}
      placement={getPlacement(props)}
      positionFixed={positionFixed}
      referenceElement={referenceElement}>
      {({ ref, ...popperProps }: PopperChildrenProps) => children({
        ...popperProps,
        innerRef: ref,
        inputHeight: referenceElement ? referenceElement.offsetHeight : 0,
      })}
    </Popper>
  );
};

Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;

export default Overlay;
