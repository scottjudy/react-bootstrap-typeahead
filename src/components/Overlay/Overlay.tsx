import PropTypes from 'prop-types';
import { CSSProperties, ReactElement, RefCallback } from 'react';

import useOverlay, { OverlayOptions, ReferenceElement } from './useOverlay';

import { ALIGN_VALUES } from '../../constants';
import { noop } from '../../utils';

// `Element` is not defined during server-side rendering, so shim it here.
/* istanbul ignore next */
const SafeElement = typeof Element === 'undefined' ? noop : Element;

const propTypes = {
  /**
   * Specify menu alignment. The default value is `justify`, which makes the
   * menu as wide as the input and truncates long values. Specifying `left`
   * or `right` will align the menu to that side and the width will be
   * determined by the length of menu item values.
   */
  align: PropTypes.oneOf(ALIGN_VALUES),
  children: PropTypes.func.isRequired,
  /**
   * Specify whether the menu should appear above the input.
   */
  dropup: PropTypes.bool,
  /**
   * Whether or not to automatically adjust the position of the menu when it
   * reaches the viewport boundaries.
   */
  flip: PropTypes.bool,
  isMenuShown: PropTypes.bool,
  positionFixed: PropTypes.bool,
  // @ts-ignore
  referenceElement: PropTypes.instanceOf(SafeElement),
};

const defaultProps = {
  align: 'justify',
  dropup: false,
  flip: false,
  isMenuShown: false,
  positionFixed: false,
};

export interface OverlayRenderProps {
  innerRef: RefCallback<HTMLElement>;
  style: CSSProperties;
}

export interface OverlayProps extends OverlayOptions {
  children: (props: OverlayRenderProps) => ReactElement | null;
  isMenuShown: boolean;
  referenceElement: ReferenceElement;
}

const Overlay = ({ referenceElement, isMenuShown, ...props }: OverlayProps) => {
  const overlayProps = useOverlay(referenceElement, props);

  if (!isMenuShown) {
    return null;
  }

  return props.children(overlayProps);
};

Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;

export default Overlay;
