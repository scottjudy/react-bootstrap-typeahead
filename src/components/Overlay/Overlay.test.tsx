import React from 'react';

import { Align, getModifiers, getPlacement } from './Overlay';
import * as stories from './Overlay.stories';

import {
  composeStories,
  generateSnapshots,
  getMenu,
  render,
  screen,
  waitFor,
} from '../../tests/helpers';

const { Default } = composeStories(stories);

interface PlacementPermutation {
  props: {
    align: Align;
    dropup: boolean;
  };
  received: string;
}

interface ModifierProps {
  align: Align;
  flip: boolean;
}

describe('<Overlay>', () => {
  generateSnapshots(stories);

  it('renders the overlay', async () => {
    render(<Default />);

    await waitFor(() => {
      expect(getMenu()).toBeInTheDocument();
    });
  });

  it('does not render the overlay when `isMenuShown=false`', () => {
    render(<Default isMenuShown={false} />);
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('updates the positioning type', async () => {
    // The story uses fixed positioning by default.
    const { rerender } = render(<Default />);
    await waitFor(() => {
      expect(getMenu()).toHaveStyle('position: fixed');
    });

    rerender(<Default positionFixed={false} />);
    await waitFor(() => {
      expect(getMenu()).toHaveStyle('position: absolute');
    });
  });
});

describe('Overlay placement', () => {
  it('computes the placement string', () => {
    const permutations: PlacementPermutation[] = [
      { props: { align: 'right', dropup: false }, received: 'bottom-end' },
      { props: { align: 'left', dropup: false }, received: 'bottom-start' },
      { props: { align: 'justify', dropup: false }, received: 'bottom-start' },
      { props: { align: 'foo', dropup: false }, received: 'bottom-start' },
      { props: { align: 'right', dropup: true }, received: 'top-end' },
      { props: { align: 'left', dropup: true }, received: 'top-start' },
      { props: { align: 'justify', dropup: true }, received: 'top-start' },
      { props: { align: 'foo', dropup: true }, received: 'top-start' },
    ];

    permutations.forEach(({ props, received }) => {
      expect(getPlacement(props)).toBe(received);
    });
  });
});

describe('Overlay modifiers', () => {
  it('sets the `flip` modifier', () => {
    const props: ModifierProps = { align: 'justify', flip: false };
    const selector = ({ name }) => name === 'flip';

    expect(getModifiers(props).find(selector).enabled).toBe(false);

    props.flip = true;
    expect(getModifiers(props).find(selector).enabled).toBe(true);
  });

  it('conditionally adds the `setWidth` modifier', () => {
    const props: ModifierProps = { align: 'justify', flip: false };

    const modifiers = getModifiers(props);
    expect(modifiers).toHaveLength(2);
    expect(
      modifiers.find(({ name }) => name === 'setPopperWidth')
    ).toBeTruthy();

    props.align = 'left';
    expect(getModifiers(props)).toHaveLength(1);

    props.align = 'right';
    expect(getModifiers(props)).toHaveLength(1);
  });
});
