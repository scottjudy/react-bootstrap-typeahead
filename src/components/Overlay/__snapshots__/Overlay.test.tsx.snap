// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay> Default story renders snapshot 1`] = `
Array [
  <div
    style={
      Object {
        "border": "1px solid red",
        "padding": "4px",
      }
    }
  >
    Reference element
  </div>,
  <div
    aria-label="menu-options"
    className="rbt-menu dropdown-menu show"
    id="overlay-story"
    onMouseDown={[Function]}
    overlayRef={[Function]}
    role="listbox"
    style={
      Object {
        "display": "block",
        "left": "0",
        "maxHeight": "300px",
        "overflow": "auto",
        "position": "fixed",
        "top": "0",
      }
    }
  >
    <a
      className="dropdown-item"
      href="#"
      onClick={[Function]}
      role="option"
    >
      This is the menu
    </a>
  </div>,
]
`;
